beacon-chain/blockchain/service.go

```diff
@@ -222,28 +226,46 @@ func (c *ChainService) blockProcessing(processedBlock chan<- *types.Block) {
 		// can be received either from the sync service, the RPC service,
 		// or via p2p.
 		case block := <-c.incomingBlockChan:
-			if err := c.processBlock(block); err != nil {
-				log.Error(err)
-				processedBlock <- nil
+
+			beaconState, err := c.beaconDB.GetState()
+			if err != nil {
+				log.Errorf("Unable to retrieve beacon state %v", err)
+			}
+
+			currentSlot := beaconState.Slot()
+
+			if currentSlot+1 < block.SlotNumber() {
+				c.unProcessedBlocks[block.SlotNumber()] = block
 				continue
 			}
 
-			// Push the block to trigger the fork choice rule.
-			processedBlock <- block
+			if currentSlot+1 == block.SlotNumber() {
+
+				if err := c.processBlock(block); err != nil {
+					log.Error(err)
+					processedBlock <- nil
+					continue
+				}
+
+				// Push the block to trigger the fork choice rule.
+				processedBlock <- block
+			} else {
+				log.Debugf(
+					"Block slot number is lower than the current slot in the beacon state %d",
+					block.SlotNumber())
+				c.sendAndDeleteCachedBlocks(currentSlot)
```

Comment:
Why do we do this? Wouldn't we hold onto blocks at a slot until a finalized checkpoint, in case of reorg?

Reply:
This is just to handle cases where a block arrives to early and we cant process it yet, The spec doesnt go into details on how we will handle reorgs or state rollbacks yet

Reply:
Sounds good, but i think it something we might want to optimize in the future